### PR TITLE
Improve error reporting

### PR DESCRIPTION
**Pull request by matthid**
This improves output and internal error propagation in the NuGet-RequestVersions code.
The goal is to output in the error case which urls where requested and what reponse we got from them...

For now this is WIP, because I think there might be some cases now where we print way too much info (in the non-verbose case).

### CODE REVIEW

File: src/Paket.Core/Common/Async.fs [hunk 1]
```diff
@@ -17,34 +18,47 @@ module AsyncExtensions =
             let! b'' = b'
             return (a'',b'')
         }
-
-    static member Choice(tasks : Async<'T option> seq) = async {
+    static member map f a =
+        async { return f a }
+    static member tryFind (f : 'T -> bool) (tasks : Async<'T> seq) = async {
```
Comment (matthid) [inline src/Paket.Core/Common/Async.fs]: This is an interesting API.
It is almost like `Async.Choice`, but will return all the started tasks and an index which one matched the given function.

File: src/Paket.Core/Common/Utils.fs [hunk 2]
```diff
@@ -560,7 +555,7 @@ let downloadFromUrl (auth:Auth option, url : string) (filePath: string) =
             do! task
         with
         | exn ->
-            failwithf "Could not download from %s%s Message: %s%s" url Environment.NewLine exn.Message (innerText exn)
+            raise <| Exception(sprintf "Could not download from '%s'" url, exn)
```
Comment (matthid) [inline src/Paket.Core/Common/Utils.fs]: We almost never should do `failwithf` in a general `with exn ->` clause, not even when we embed the message (like here)...
I really think F# should make this more easy (with some syntax), even the F# compiler is doing it wrong everywhere

/cc @dsyme any ideas?
maybe allowing to use comma in combination with `failwithf "some message", innerException`

File: src/Paket.Core/Common/Utils.fs [hunk 3]
```diff
@@ -612,11 +606,12 @@ let safeGetFromUrl (auth:Auth option, url : string, contentType : string) =
 #endif
             use _ = Profile.startCategory Profile.Category.NuGetRequest
             let! raw = client.DownloadStringTaskAsync(uri) |> Async.AwaitTask
-            return Some raw
+            return FSharp.Core.Result.Ok raw
```
Comment (matthid) [inline src/Paket.Core/Common/Utils.fs]: Maybe I should use full-out exceptions everywhere :/

**Comment by matthid** [comment]
This is how a failure looks now:

```
Paket failed with:
-> AggregateException: Could not find versions for package Argu on https://www.myget.org/F/paket-test.
-> AggregateException: Source 'https://www.myget.org/F/paket-test' yielded no results
	 - Request 'https://www.myget.org/F/paket-test/FindPackagesById()?semVerLevel=2.0.0&id='Argu'' finished with error
	-> Could not retrieve data from 'https://www.myget.org/F/paket-test/FindPackagesById()?semVerLevel=2.0.0&id='Argu''
	-> WebException: The remote server returned an error: (401) Unauthorized.
	 - Request 'https://www.myget.org/F/paket-test/Packages?semVerLevel=2.0.0&$filter=tolower(Id) eq 'argu'' finished with error
	-> Could not retrieve data from 'https://www.myget.org/F/paket-test/Packages?semVerLevel=2.0.0&$filter=tolower(Id) eq 'argu''
	-> WebException: The remote server returned an error: (401) Unauthorized.
```

These changes also show that the random failure we get sometimes is apparently another mono bug:
```
Paket failed with:
-> AggregateException: Could not find versions for package WebSharper on https://www.nuget.org/api/v2.
-> AggregateException: Source 'https://www.nuget.org/api/v2' yielded no results
	 - Request 'https://www.nuget.org/api/v2/FindPackagesById()?semVerLevel=2.0.0&id='WebSharper'' finished with error
	-> Could not retrieve data from 'https://www.nuget.org/api/v2/FindPackagesById()?semVerLevel=2.0.0&id='WebSharper''
	-> WebException: Error: SecureChannelFailure (Ssl error:100000f7:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER
	     at /tmp/buildd/mono-5.0.1.1/external/boringssl/ssl/tls_record.c:217)
	-> MonoBtlsException: Ssl error:100000f7:SSL routines:OPENSSL_internal:WRONG_VERSION_NUMBER
	     at /tmp/buildd/mono-5.0.1.1/external/boringssl/ssl/tls_record.c:217
	 - Request 'https://www.nuget.org/api/v2/Packages?semVerLevel=2.0.0&$filter=tolower(Id) eq 'websharper'' finished with error
	-> Could not retrieve data from 'https://www.nuget.org/api/v2/Packages?semVerLevel=2.0.0&$filter=tolower(Id) eq 'websharper''
	-> WebException: Error: TrustFailure (Ssl error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
	     at /tmp/buildd/mono-5.0.1.1/external/boringssl/ssl/handshake_client.c:1132)
	-> MonoBtlsException: Ssl error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
	     at /tmp/buildd/mono-5.0.1.1/external/boringssl/ssl/handshake_client.c:1132
```

**Comment by matthid** [comment]
https://github.com/mono/boringssl/blob/mono/ssl/s3_both.c#L363

https://github.com/mono/boringssl/blob/mono/ssl/s3_both.c#L291

> The record length is a 16-byte value and is formatted in network order.

> In theory, this means that a single record can be up to 65,535 (2^16 -1) bytes in length. The TLSv1 RFC2246 states that the maximum length is 16,383 (2^14 -1) bytes. Microsoft products (Microsoft Internet Explorer and Internet Information Services) are known to exceed these limits.

http://www.cisco.com/c/en/us/support/docs/security-vpn/secure-socket-layer-ssl/116181-technote-product-00.html

Nice

**Comment by matthid** [comment]
Note that

```
1) Error : Paket.IntegrationTests.ConvertFromNuGetSpecs.#1217 should replace packages.config files in project
System.Exception : Paket failed with:
-> Exception: There was a version conflict during package resolution.
     Conflict detected:
      - Dependencies file requested package Newtonsoft.Json: 7.0.1
      - Available versions:
        - (10.0.2, [https://www.nuget.org/api/v2])
       ... snip
        - (8.0.1-beta1, [https://www.nuget.org/api/v2])
        - (7.0.1, [https://dotnet.myget.org/F/cli-deps; https://www.nuget.org/api/v2])
        - (7.0.1-beta3, [https://www.nuget.org/api/v2])
        - (7.0.1-beta2, [https://www.nuget.org/api/v2])
        - (7.0.1-beta1, [https://www.nuget.org/api/v2])
        - (6.0.8, [https://www.nuget.org/api/v2])
       ... snip
        - (3.5.8, [https://www.nuget.org/api/v2])
   
     Please try to relax some conditions or resolve the conflict manually (see http://fsprojects.github.ioPaket/nuget-dependencies.html#Use-exactly-this-version-constraint).
```

Is explained by the mono bug as well:
I think the request to `GetPackageDetails` was failing.
Paket needs to improve this error message to include this (important) detail...

**Comment by matthid** [comment]
There are so many more places where `with _ ->` is used throughout the paket code base, but I'm happy with this subset for now as especially in the `NuGet` request logic it was often hard to figure out what was going on.

**Comment by forki** [comment]
is this ready?

**Comment by forki** [comment]
thx!

**Pull request merged**